src/protocol/prf_sharding/feature_label_dot_product.rs

```diff
@@ -0,0 +1,494 @@
+        let num_user_rows = rows_for_user.len();
+        let contexts = ctx_for_row_number[..num_user_rows - 1].to_owned();
+        let record_ids = record_id_for_row_depth[..num_user_rows].to_owned();
+
+        for count in record_id_for_row_depth.iter_mut().take(rows_for_user.len()) {
+            *count += 1;
+        }
+        async move { evaluate_per_user_attribution_circuit(contexts, record_ids, rows_for_user) }
+    }));
+
+    // Execute all of the async futures (sequentially), and flatten the result
+    let flattenned_stream = seq_join(sh_ctx.active_work(), stream_of_per_user_circuits)
+        .map(|x| x.unwrap().into_iter())
+        .flatten_iters();
+
+    // modulus convert feature vector bits from shares in `Z_2` to shares in `Z_p`
+    let converted_feature_vector_bits = convert_bits(
+        prime_field_ctx
+            .narrow(&Step::ModulusConvertFeatureVectorBits)
+            .set_total_records(num_outputs),
+        flattenned_stream,
+        0..FV::BITS,
+    );
+
+    // Sum up all the vectors
+    converted_feature_vector_bits
+        .try_fold(
+            vec![S::ZERO; usize::try_from(FV::BITS).unwrap()],
+            |mut running_sums, row_contribution| async move {
+                for (i, contribution) in row_contribution.iter().enumerate() {
+                    running_sums[i] += contribution;
+                }
+                Ok(running_sums)
+            },
+        )
+        .await
+}
+
+async fn evaluate_per_user_attribution_circuit<C, FV>(
+    ctx_for_row_number: Vec<C>,
+    record_id_for_each_depth: Vec<u32>,
+    rows_for_user: Vec<PrfShardedIpaInputRow<FV>>,
+) -> Result<Vec<BitDecomposed<Replicated<Gf2>>>, Error>
+where
+    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    FV: GaloisField,
+{
+    assert!(!rows_for_user.is_empty());
+    if rows_for_user.len() == 1 {
+        return Ok(Vec::new());
+    }
+    let first_row = &rows_for_user[0];
+    let mut prev_row_inputs = initialize_new_device_attribution_variables(first_row);
+
+    let mut output = Vec::with_capacity(rows_for_user.len() - 1);
+    // skip the first row as it requires no multiplications
+    // no context was created for the first row
+    for (i, (row, ctx)) in
+        zip(rows_for_user.iter().skip(1), ctx_for_row_number.into_iter()).enumerate()
+    {
+        let record_id_for_this_row_depth = RecordId(record_id_for_each_depth[i + 1]); // skip row 0
+
+        let capped_attribution_outputs = prev_row_inputs
+            .compute_row_with_previous(ctx, record_id_for_this_row_depth, row)
+            .await?;
+
+        output.push(capped_attribution_outputs);
+    }
+
+    Ok(output)
+}
+
+///
+/// Upon encountering the first row of data from a new user (as distinguished by a different OPRF of the match key)
+/// this function encapsulates the variables that must be initialized. No communication is required for this first row.
+///
+fn initialize_new_device_attribution_variables<FV>(
```

Comment:
this is now just a `impl From<&PrfShardedIpaInputRow> for InputsRequiredFromPrevRow` maybe?

Reply:
I can't figure out how to do this.